### PR TITLE
WIP: keepalive worker ssh connections

### PR DESCRIPTION
**Pull request by vito**
note: i haven't tested this out with a failing worker yet. it would be nice to have topgun coverage for this sort of thing where we can simulate a hard shutdown. maybe we should try it out before merging?

i also snuck in a cleanup of the `.Accept` loop which is technically unrelated but the bare `return` on any error felt a bit scary in light of possible 'max open files' errors (which would previously cause the TSA to exit).

for the future: we should probably move to https://github.com/gliderlabs/ssh when we have time. once we're using that we can set idle timeout instead, which feels a *little* more intuitive (this is the only place where we have the *server* set tcpkeepalives, usually it's done via the dialer). an idle timeout should be sufficient since we can set it based on the heartbeat interval, which should keep the connection alive.

### CODE REVIEW

File: tsa/tsacmd/server.go [hunk 1]
```diff
@@ -83,22 +83,57 @@ func (forward ForwardedTCPIP) Wait() {
...
 		if err != nil {
-			if !strings.Contains(err.Error(), "use of closed network connection") {
-				server.logger.Error("failed-to-accept", err)
+			if strings.Contains(err.Error(), "use of closed network connection") {
+				// shutting down
+				return nil
 			}
 
-			return
+			if ne, ok := err.(net.Error); ok && ne.Temporary() {
```
Comment (cirocosta) [inline tsa/tsacmd/server.go]: That's interesting, I didn't know what that had to do with the comment around the number of open files

```go
func (e Errno) Temporary() bool {
	return e == EINTR || e == EMFILE || e.Timeout()
}

//        EMFILE The per-process limit on the number of open file descriptors has been reached.
```

([syscall_unix.go](https://github.com/golang/go/blob/ccd9d9d4cef16618d84bae9cba9f1b264278c8ab/src/syscall/syscall_unix.go#L123-L125))

Neat!

File: tsa/tsacmd/server.go [hunk 2]
```diff
@@ -83,22 +83,57 @@ func (forward ForwardedTCPIP) Wait() {
...
 		})
 
-		go server.handshake(logger, c)
+		if tc, ok := conn.(*net.TCPConn); ok {
+			err := tc.SetKeepAlive(true)
+			if err != nil {
+				logger.Error("failed-to-set-tcp-keepalive", err)
+			}
+
+			err = tc.SetKeepAlivePeriod(15 * time.Second)
```
Comment (cirocosta) [inline tsa/tsacmd/server.go]: Despite having looked at the other PR regarding keepalives yesterday, I got wondering about the number of probes before considering the connection idle. 

For those who might end up like me, check out @vito's commit message here https://github.com/concourse/concourse/commit/cac9c6d4435701757a96fb4f2a871cb4180d17d8 . tl;dr:  going with the default as Go doesn't expose a standard way of setting the number of probes.

File: tsa/tsacmd/server_runner.go [hunk 3]
```diff
@@ -26,17 +26,16 @@ func (runner serverRunner) Run(signals <-chan os.Signal, ready chan<- struct{})
...
-		runner.server.Serve(listener)
+		exited <- runner.server.Serve(listener)
 	}()
 
 	for {
 		select {
-		case <-exited:
-			return nil
+		case err := <-exited:
+			return err
```
Comment (cirocosta) [inline tsa/tsacmd/server_runner.go]: :+1:

**Comment by vito** [comment]
@cirocosta Yeah, I think we could even write a `topgun` test for this that partitions the worker's network. I don't know what to assert exactly at this point, but the pipeline should eventually 'recover' and start using a different, healthy worker.

**Pull request closed without merging**